File: python/advanced/avl_tree.py
"""
AVL Tree (Adelson-Velsky and Landis)
=====================================
A self-balancing Binary Search Tree where the heights of the two child
subtrees of any node differ by at most one. Rebalancing is done via single
and double rotations after every insert/delete.

Complexities:
    - Insert:  O(log n)   — guaranteed, not amortized
    - Delete:  O(log n)
    - Search:  O(log n)
    - Space:   O(n)
"""

from __future__ import annotations
from typing import Optional, Any, List


class AVLNode:
    """A node that also stores its subtree height for O(1) balance queries."""

    def __init__(self, key: Any) -> None:
        self.key: Any = key
        self.left: Optional[AVLNode] = None
        self.right: Optional[AVLNode] = None
        self.height: int = 1  # new node is a leaf

    def __repr__(self) -> str:
        return f"AVLNode({self.key}, h={self.height})"


class AVLTree:
    """
    AVL Tree with insert, delete, search, and ASCII visualization.

    All mutation methods return (or internally re-assign) the subtree root so
    the tree re-links itself after rotations without a parent pointer.
    """

    def __init__(self) -> None:
        self._root: Optional[AVLNode] = None

    # ------------------------------------------------------------------
    # Height helper
    # ------------------------------------------------------------------

    @staticmethod
    def _h(node: Optional[AVLNode]) -> int:
        """Return stored height, or 0 for None."""
        return node.height if node else 0

    def _update_height(self, node: AVLNode) -> None:
        node.height = 1 + max(self._h(node.left), self._h(node.right))

    # ------------------------------------------------------------------
    # Balance factor
    # ------------------------------------------------------------------

    def get_balance(self, node: Optional[AVLNode]) -> int:
        """
        Balance factor = height(left) - height(right).
        A value outside [-1, 1] means the node is unbalanced.

        Time:  O(1)
        """
        if node is None:
            return 0
        return self._h(node.left) - self._h(node.right)

    # ------------------------------------------------------------------
    # Rotations
    # ------------------------------------------------------------------

    def rotate_right(self, z: AVLNode) -> AVLNode:
        """
        Right rotation around *z*.

              z                y
             / \\             / \\
            y   T4    →    x    z
           / \\                 / \\
          x   T3              T3  T4

        Time:  O(1)
        """
        y = z.left
        T3 = y.right

        y.right = z
        z.left = T3

        self._update_height(z)
        self._update_height(y)
        return y  # new subtree root

    def rotate_left(self, z: AVLNode) -> AVLNode:
        """
        Left rotation around *z*.

          z                   y
         / \\                 / \\
        T1   y      →      z    x
            / \\           / \\
           T2   x         T1  T2

        Time:  O(1)
        """
        y = z.right
        T2 = y.left

        y.left = z
        z.right = T2

        self._update_height(z)
        self._update_height(y)
        return y  # new subtree root

    # ------------------------------------------------------------------
    # Rebalance
    # ------------------------------------------------------------------

    def _rebalance(self, node: AVLNode) -> AVLNode:
        """Apply the appropriate rotation(s) if *node* is unbalanced."""
        self._update_height(node)
        balance = self.get_balance(node)

        # Left-Left case
        if balance > 1 and self.get_balance(node.left) >= 0:
            return self.rotate_right(node)

        # Left-Right case
        if balance > 1 and self.get_balance(node.left) < 0:
            node.left = self.rotate_left(node.left)
            return self.rotate_right(node)

        # Right-Right case
        if balance < -1 and self.get_balance(node.right) <= 0:
            return self.rotate_left(node)

        # Right-Left case
        if balance < -1 and self.get_balance(node.right) > 0:
            node.right = self.rotate_right(node.right)
            return self.rotate_left(node)

        return node  # already balanced

    # ------------------------------------------------------------------
    # Insert
    # ------------------------------------------------------------------

    def insert(self, key: Any) -> None:
        """
        Insert *key* and rebalance along the path back to the root.

        Time:  O(log n)
        Space: O(log n) call stack
        """
        self._root = self._insert(self._root, key)

    def _insert(self, node: Optional[AVLNode], key: Any) -> AVLNode:
        if node is None:
            return AVLNode(key)
        if key < node.key:
            node.left = self._insert(node.left, key)
        elif key > node.key:
            node.right = self._insert(node.right, key)
        else:
            return node  # duplicate — no change
        return self._rebalance(node)

    # ------------------------------------------------------------------
    # Delete
    # ------------------------------------------------------------------

    def delete(self, key: Any) -> None:
        """
        Delete *key* and rebalance.

        Time:  O(log n)
        Space: O(log n) call stack
        """
        self._root = self._delete(self._root, key)

    def _delete(self, node: Optional[AVLNode], key: Any) -> Optional[AVLNode]:
        if node is None:
            return None
        if key < node.key:
            node.left = self._delete(node.left, key)
        elif key > node.key:
            node.right = self._delete(node.right, key)
        else:
            if node.left is None:
                return node.right
            if node.right is None:
                return node.left
            # Two children: replace with in-order successor.
            successor = self._min_node(node.right)
            node.key = successor.key
            node.right = self._delete(node.right, successor.key)
        return self._rebalance(node)

    def _min_node(self, node: AVLNode) -> AVLNode:
        while node.left:
            node = node.left
        return node

    # ------------------------------------------------------------------
    # Search
    # ------------------------------------------------------------------

    def search(self, key: Any) -> bool:
        """
        Return True if *key* is in the tree.

        Time:  O(log n)
        Space: O(1) iterative
        """
        node = self._root
        while node:
            if key == node.key:
                return True
            node = node.left if key < node.key else node.right
        return False

    # ------------------------------------------------------------------
    # Height (public)
    # ------------------------------------------------------------------

    def height(self) -> int:
        """Return the height of the tree. Empty tree → 0."""
        return self._h(self._root)

    # ------------------------------------------------------------------
    # ASCII visualization
    # ------------------------------------------------------------------

    def __str__(self) -> str:
        """Sideways ASCII print, right subtree on top."""
        lines: List[str] = []
        self._build_str(self._root, 0, lines)
        return "\n".join(lines) if lines else "<empty>"

    def _build_str(
        self, node: Optional[AVLNode], level: int, lines: List[str]
    ) -> None:
        if node is None:
            return
        self._build_str(node.right, level + 1, lines)
        lines.append("  " * level + f"{node.key}(h={node.height})")
        self._build_str(node.left, level + 1, lines)


# ----------------------------------------------------------------------
# Demo
# ----------------------------------------------------------------------

if __name__ == "__main__":
    avl = AVLTree()
    keys = [10, 20, 30, 40, 50, 25]
    print("=== AVL Tree ===")
    print(f"Inserting: {keys}")
    for k in keys:
        avl.insert(k)

    print(avl)
    print()
    print("Height    :", avl.height())
    print("Balance(root):", avl.get_balance(avl._root))
    print()
    print("Search 25 :", avl.search(25))
    print("Search 99 :", avl.search(99))
    print()
    avl.delete(40)
    print("After deleting 40:")
    print(avl)

    # Demonstrate LL, LR, RL, RR cases by sequential inserts.
    print("\n--- Inserting 1..7 (forces all rotation types) ---")
    avl2 = AVLTree()
    for k in range(1, 8):
        avl2.insert(k)
    print(avl2)
    print("Height:", avl2.height(), "(should be 3 for 7 nodes)")

File: java/advanced/AVLTree.java
package advanced;

/**
 * Self-balancing AVL Tree with {@code int} keys.
 *
 * <p>An AVL tree maintains the <em>height-balance property</em>: for every node the
 * difference between the heights of its left and right subtrees (the balance factor)
 * is at most 1.  Rotations (LL, RR, LR, RL) restore the property after mutations.
 *
 * <p>Time complexities:
 * <ul>
 *   <li>insert  – O(log n)</li>
 *   <li>delete  – O(log n)</li>
 *   <li>search  – O(log n)</li>
 *   <li>height  – O(1) (cached per node)</li>
 * </ul>
 *
 * <p>Space complexity: O(n) for the tree; O(log n) auxiliary for recursive calls.
 */
public class AVLTree {

    // -------------------------------------------------------------------------
    // Inner Node
    // -------------------------------------------------------------------------

    private static class Node {
        int key;
        int height;   // height of subtree rooted at this node (leaf = 0)
        Node left, right;

        Node(int key) {
            this.key = key;
            this.height = 0;
        }
    }

    // -------------------------------------------------------------------------
    // Fields
    // -------------------------------------------------------------------------

    private Node root;
    private int size;

    // -------------------------------------------------------------------------
    // Height helpers
    // -------------------------------------------------------------------------

    /**
     * Returns the cached height of {@code node}, or -1 for null.
     *
     * <p>Time: O(1).
     *
     * @param node the node (may be null)
     * @return height
     */
    public int height(Node node) {
        return node == null ? -1 : node.height;
    }

    /** Returns the height of the root (whole tree). */
    public int height() {
        return height(root);
    }

    private void updateHeight(Node node) {
        node.height = 1 + Math.max(height(node.left), height(node.right));
    }

    // -------------------------------------------------------------------------
    // Balance factor
    // -------------------------------------------------------------------------

    /**
     * Returns the balance factor of {@code node}: height(left) - height(right).
     *
     * <p>Time: O(1).
     *
     * @param node the node
     * @return balance factor in [-2, 2] during rebalancing
     */
    public int getBalance(Node node) {
        return node == null ? 0 : height(node.left) - height(node.right);
    }

    // -------------------------------------------------------------------------
    // Rotations
    // -------------------------------------------------------------------------

    /**
     * Performs a <em>left rotation</em> around {@code y}.
     *
     * <pre>
     *    y                  x
     *   / \               /   \
     *  T1   x    -->    y      xR
     *      / \         / \
     *    xL   xR      T1  xL
     * </pre>
     *
     * <p>Time: O(1).
     *
     * @param y the pivot node
     * @return the new root of this subtree
     */
    public Node rotateLeft(Node y) {
        Node x  = y.right;
        Node xL = x.left;

        x.left  = y;
        y.right = xL;

        updateHeight(y);
        updateHeight(x);

        return x;
    }

    /**
     * Performs a <em>right rotation</em> around {@code y}.
     *
     * <pre>
     *      y              x
     *     / \           /   \
     *    x   T3  -->  xL     y
     *   / \                 / \
     * xL   xR             xR   T3
     * </pre>
     *
     * <p>Time: O(1).
     *
     * @param y the pivot node
     * @return the new root of this subtree
     */
    public Node rotateRight(Node y) {
        Node x  = y.left;
        Node xR = x.right;

        x.right = y;
        y.left  = xR;

        updateHeight(y);
        updateHeight(x);

        return x;
    }

    // -------------------------------------------------------------------------
    // Rebalance
    // -------------------------------------------------------------------------

    private Node rebalance(Node node) {
        updateHeight(node);
        int balance = getBalance(node);

        // Left Heavy
        if (balance > 1) {
            if (getBalance(node.left) < 0) {          // LR case
                node.left = rotateLeft(node.left);
            }
            return rotateRight(node);                  // LL case
        }

        // Right Heavy
        if (balance < -1) {
            if (getBalance(node.right) > 0) {         // RL case
                node.right = rotateRight(node.right);
            }
            return rotateLeft(node);                   // RR case
        }

        return node; // already balanced
    }

    // -------------------------------------------------------------------------
    // Insert
    // -------------------------------------------------------------------------

    /**
     * Inserts {@code key} into the AVL tree.  Duplicate keys are ignored.
     *
     * <p>Time: O(log n) | Space: O(log n).
     *
     * @param key the key to insert
     */
    public void insert(int key) {
        root = insertRec(root, key);
    }

    private Node insertRec(Node node, int key) {
        if (node == null) {
            size++;
            return new Node(key);
        }
        if (key < node.key)      node.left  = insertRec(node.left,  key);
        else if (key > node.key) node.right = insertRec(node.right, key);
        else return node; // duplicate

        return rebalance(node);
    }

    // -------------------------------------------------------------------------
    // Delete
    // -------------------------------------------------------------------------

    /**
     * Deletes {@code key} from the AVL tree, rebalancing as necessary.
     *
     * <p>Time: O(log n) | Space: O(log n).
     *
     * @param key the key to delete
     */
    public void delete(int key) {
        root = deleteRec(root, key);
    }

    private Node deleteRec(Node node, int key) {
        if (node == null) return null;

        if (key < node.key) {
            node.left = deleteRec(node.left, key);
        } else if (key > node.key) {
            node.right = deleteRec(node.right, key);
        } else {
            size--;
            // Case 1 & 2: zero or one child
            if (node.left == null) return node.right;
            if (node.right == null) return node.left;
            // Case 3: two children – replace with in-order successor
            Node successor = findMin(node.right);
            node.key = successor.key;
            size++; // deleteRec will decrement again
            node.right = deleteRec(node.right, successor.key);
        }
        return rebalance(node);
    }

    private Node findMin(Node node) {
        while (node.left != null) node = node.left;
        return node;
    }

    // -------------------------------------------------------------------------
    // Search
    // -------------------------------------------------------------------------

    /**
     * Returns {@code true} if {@code key} is present in the AVL tree.
     *
     * <p>Time: O(log n) | Space: O(1).
     *
     * @param key the key to search for
     * @return {@code true} if found
     */
    public boolean search(int key) {
        Node cur = root;
        while (cur != null) {
            if      (key < cur.key) cur = cur.left;
            else if (key > cur.key) cur = cur.right;
            else return true;
        }
        return false;
    }

    // -------------------------------------------------------------------------
    // Size
    // -------------------------------------------------------------------------

    /**
     * Returns the number of keys in the AVL tree.
     *
     * <p>Time: O(1).
     *
     * @return size
     */
    public int size() { return size; }

    // -------------------------------------------------------------------------
    // ASCII tree print (toString)
    // -------------------------------------------------------------------------

    /**
     * Returns a multi-line ASCII representation showing keys and balance factors.
     *
     * <p>Time: O(n) | Space: O(n).
     *
     * @return ASCII tree string
     */
    @Override
    public String toString() {
        if (root == null) return "(empty tree)";
        StringBuilder sb = new StringBuilder();
        buildAscii(sb, root, "", "");
        return sb.toString();
    }

    private void buildAscii(StringBuilder sb, Node node, String prefix, String childPrefix) {
        if (node == null) return;
        sb.append(prefix)
          .append(node.key)
          .append(" (h=").append(node.height)
          .append(", bf=").append(getBalance(node))
          .append(")\n");
        if (node.left != null || node.right != null) {
            buildAscii(sb, node.right, childPrefix + "├── R: ", childPrefix + "│   ");
            buildAscii(sb, node.left,  childPrefix + "└── L: ", childPrefix + "    ");
        }
    }

    // -------------------------------------------------------------------------
    // Main – demo
    // -------------------------------------------------------------------------

    public static void main(String[] args) {
        AVLTree avl = new AVLTree();

        System.out.println("=== AVL Tree Demo ===\n");

        // Insert – triggers various rotations
        int[] keys = {10, 20, 30, 40, 50, 25};
        for (int k : keys) {
            avl.insert(k);
            System.out.println("After insert(" + k + "):");
            System.out.println(avl);
        }

        System.out.println("Size  : " + avl.size());
        System.out.println("Height: " + avl.height());

        // Search
        System.out.println("\nSearch 25 : " + avl.search(25));
        System.out.println("Search 99 : " + avl.search(99));

        // Delete
        System.out.println("\nDelete 10:");
        avl.delete(10);
        System.out.println(avl);

        System.out.println("Delete 30:");
        avl.delete(30);
        System.out.println(avl);

        System.out.println("Delete 40 (two children case):");
        avl.delete(40);
        System.out.println(avl);

        // Stress test – insert descending (would degenerate in plain BST)
        AVLTree avl2 = new AVLTree();
        System.out.println("Descending insert 20..1 – height should be ~4:");
        for (int i = 20; i >= 1; i--) avl2.insert(i);
        System.out.println("Height: " + avl2.height() + "  (log2(20) ≈ 4.3)");
        System.out.println(avl2);
    }
}
